add keycloak type to configure function

laurafitzgerald opened the pull request:
# Description
https://issues.redhat.com/browse/MGDAPI-1227

## Verification Steps

### Install the operator

By running the following commands
```
INSTALLATION_TYPE=managed-api make cluster/prepare/local
INSTALLATION_TYPE=managed-api IN_PROW=true make deploy/integreatly-rhmi-cr.yml
INSTALLATION_TYPE=managed-api make code/run
```

During the first installation verify that the operator has been provided with a secret 1

`oc get rhmi rhoam -n redhat-rhoam-operator -o yaml | yq - r 'status.toSKU'` -> 1

Wait for the installation to fully complete. You can verify this by 
`oc get rhmi rhoam -n redhat-rhoam-operator -o yaml | yq - r 'status.toSKU'` -> nil

### Verify that the keycloak cr has the expected values

By running the following commands
```
oc get keycloak rhssouser -n redhat-rhoam-user-sso -o yaml | yq - r 'spec.keycloakDeploymentSpec.resources'
```
output
```
  resources:
      limits:
        cpu: 700m
        memory: 1k
      requests:
        cpu: 700m
        memory: 1k

```

`oc get keycloak rhssouser -n redhat-rhoam-user-sso | yq - r 'spec.replicas' ` -> 1


### Verify that after changing the values lower, they will get updated on the next reconcile

Change the resource block via the UI to all lower values 

Wait for the next reconcile to complete and recheck the values using the commands and output from previous step.

### Verify that changing the SKU secret value will propagate 

## Type of change

<!-- Please delete options that are not relevant. -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist
- [ ] This change requires a documentation update <!-- Update JIRA with Affects -> Documentation -->
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added a test case that will be used to verify my changes 
- [ ] Verified independently on a cluster by reviewer

laurafitzgerald commented on pkg/products/rhssouser/reconciler.go:
```diff
@@ -274,14 +273,27 @@ func (r *Reconciler) Reconcile(ctx context.Context, installation *integreatlyv1a
 			Namespace: r.Config.GetNamespace(),
 		},
 	}
+
```
Note. This change needs to be tied to managed-api.

KevFan commented on pkg/resources/sku/sku.go:
```diff
@@ -159,12 +161,18 @@ func (p AProductConfig) Configure(obj metav1.Object) error {
+			t.Spec.Instances = int(configReplicas)
+		}
+		resources := p.resourceConfigs[KeycloakName].Resources
+		p.mutateResources(t.Spec.KeycloakDeploymentSpec.Resources.Requests, resources.Requests)
+		p.mutateResources(t.Spec.KeycloakDeploymentSpec.Resources.Limits, resources.Limits)
```
Doing a fresh install, `t.Spec.KeycloakDeploymentSpec.Resources.Requests` seems to be nil here ? :thinking:

KevFan commented on pkg/resources/sku/sku.go:
```diff
@@ -190,15 +198,16 @@ func (p AProductConfig) mutate(podTemplateSpec *v13.PodTemplateSpec, name string
 	if p.sku.isUpdated || podcpu.Cmp(cfg[v13.ResourceCPU]) == -1 || podcpu.IsZero() {
-		pod[v13.ResourceCPU] = cfg[v13.ResourceCPU]
+		quantity := cfg[v13.ResourceCPU]
+		pod[v13.ResourceCPU] = resource.MustParse(quantity.String())
```
Getting a nil pointer here on a fresh install possibly due to pod can be nil :thinking:

KevFan reviewed (commented):
Overall looks good :+1: . Left some comments, hitting some issue with install for verification :thinking:

laurafitzgerald commented:
@KevFan cheers for the review. I've updated the keycloak case to ensure the blocks are not empty. I've also added a test case to cover that. I'm testing it now. 👍

KevFan reviewed (commented):
Nice work :+1:  Left some comments / questions :man_technologist:

laurafitzgerald commented on pkg/resources/sku/sku.go:
```diff
@@ -178,27 +190,32 @@ func (p AProductConfig) mutate(podTemplateSpec *v13.PodTemplateSpec, name string
 	}
 }
 
 func (p AProductConfig) mutateResources(pod, cfg v13.ResourceList) {
```
The go syntax allows more than one function paramater to be defined of which they'll take the next defined type rather than redifining the type more than once.

laurafitzgerald commented on pkg/resources/sku/sku.go:
```diff
@@ -178,27 +190,32 @@ func (p AProductConfig) mutate(podTemplateSpec *v13.PodTemplateSpec, name string
-		if container.Resources.Limits == nil {
-			podTemplateSpec.Spec.Containers[i].Resources.Limits = make(map[corev1.ResourceName]resource.Quantity)
-		}
+		checkResourceBlock(&container.Resources)
 		p.mutateResources(podTemplateSpec.Spec.Containers[i].Resources.Limits, resources.Limits)
```
👍  me and points. 🙈  will fix this up. 👍

laurafitzgerald commented on pkg/resources/sku/sku.go:
```diff
@@ -178,27 +190,32 @@ func (p AProductConfig) mutate(podTemplateSpec *v13.PodTemplateSpec, name string
 	//Cmp returns -1 if the quantity is less than y (passed value) so if podmem is less than cfg memory
 	if p.sku.isUpdated || podmem.Cmp(cfg[v13.ResourceMemory]) == -1 || podmem.IsZero() {
-		pod[v13.ResourceMemory] = cfg[v13.ResourceMemory]
+		quantity := cfg[v13.ResourceMemory]
+		pod[v13.ResourceMemory] = resource.MustParse(quantity.String())
```
for sure. 👍

laurafitzgerald commented on pkg/resources/sku/sku.go:
```diff
@@ -159,12 +161,22 @@ func (p AProductConfig) Configure(obj metav1.Object) error {
 	}
 
-	configReplicas := p.resourceConfigs[obj.GetName()].Replicas
 	if p.sku.isUpdated || *replicas < configReplicas {
 		*replicas = configReplicas
 	}
```
yep @makslion also spotted this one, I'll fix that up. 👍

laurafitzgerald commented:
Hi @KevFan this one is good for another reivew and verification if you have the time. 👍

KevFan commented:
Code looks good to me and verified the following:

* Verified `toSKU` is `1` during install on RHOAM CR
* Verified `toSKU` is `nil` after install on RHOAM CR
* Verified User SSO `KeyCloak` CR has expected values
![image](https://user-images.githubusercontent.com/24636860/114202882-8bb08680-994f-11eb-9dbd-10310c46678d.png)
* Verified replicas
![image](https://user-images.githubusercontent.com/24636860/114202974-a4b93780-994f-11eb-9434-50c3e17465c0.png)
* Verified changing resource values to lower value in User SSO `Keycloak` CR will be reverted on next reconcile
* Verified changing SKU secret value will propagate changes to User SSO `Keycloak CR`
![image](https://user-images.githubusercontent.com/24636860/114205063-d0d5b800-9951-11eb-8c82-7b2c667ac796.png)

Nice work :+1: 
/lgtm
